### PR TITLE
Add timer and PWM

### PR DESCRIPTION
**Pull request by ketotek**
I've implemented timer and PWM support and I thought others might find it useful. This is my first real HAL implementation, so let me know how to improve it, if needed. I took inspiration from various other HAL crates. I'm not really happy with the design of the `PwmTimer` constructor: it tries to capture references to the output pins used by the timer channel, but you can still make mistakes by mixing up the order.

I've tested the first 2-3 timers and theirs channels on real hardware and verified that duty cycle and frequency are correct. However, there are still some corner cases that I didn't look into, like timers with a single channel.

### CODE REVIEW

**Comment by Disasm** [comment]
Thank you! I will try to review this in a few days

File: src/pwm.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+                        duty: [0f32; 4],
+                        ch0,
+                        ch1,
+                        ch2,
+                        ch3,
+                    };
+
+                    apb.en().modify(|_r, w| w.$timXen().set_bit());
+                    apb.rstr().modify(|_r, w| w.$timXrst().set_bit());
+                    apb.rstr().modify(|_r, w| w.$timXrst().clear_bit());
```
Comment (Disasm) [inline src/pwm.rs]: It's better to use `$TIM::enable();` and `$TIM::reset();` here

File: src/pwm.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+                    timer
+                }
+            }
+
+            impl<'a> Pwm for PwmTimer<'a, $TIM> {
+                type Channel = Channel;
+                type Time = Hertz;
+                type Duty = f32;
```
Comment (Disasm) [inline src/pwm.rs]: I think that integer type is more suitable here

File: src/pwm.rs (hunk 1 above)
Comment (ketotek) [reply src/pwm.rs]: Thanks, I missed those traits. Any idea why they are not implemented for TIMER4-TIMER6? I can add that too, but I wanted to understand if there is some limitation about those timers.

Comment (Disasm) [reply src/pwm.rs]: I think that I forgot to add those.

File: src/pwm.rs (hunk 2 above)
Comment (ketotek) [reply src/pwm.rs]: I was also not sure about this, I only wanted a way to express the full duty cycle independently of the timer configuration. But I guess the simplest would be to use the integer values and let the caller deal with that.

**Comment by Disasm** [comment]
Published v0.1.0

File: src/pwm.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                }
+
+                fn set_period<P>(&mut self, period: P) where
+                    P: Into<Self::Time> {
+                    self.timer.ctl0.modify(|_, w| w.cen().clear_bit());
+                    self.timer.cnt.reset();
+
+                    let freq = period.into();
+
+                    let timer_clock = self.clocks.sysclk().0;
```
Comment (Disasm) [inline src/pwm.rs]: Hmm. Seems like timers use wrong frequency source. I need to check this on my hardware.

Comment (ketotek) [reply src/pwm.rs]: You're right, I missed that part from the datasheet and didn't realize TIMER0 has a different clock too. I'll send a PR, it's a 3 line change to select the correct clock.

**Pull request merged**